---
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    highlight: zenburn
    theme: flatly
---

```{r}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

```{r}
library(ggplot2)
library(dplyr)
library(tidyr)
library(biomaRt)
library(pheatmap)
```

# Overview
This is looking at a first sequencing run of single cells.
There are three samples, `H1`, `H2` and `H3`. There is some prior knowledge that
the `H1` sample might have some quality problems.

## Barcode distributions {.tabset}

For a single-cell experiment, we expect to see a set of barcodes that come
from real cells and then a drop. Below we show the barcode distribution
for a published experiment on K562 cells and then each of these samples.
We look at only the top 100,000 barcodes for each sample.

If you click on the tabs you can see each of the charts. We can see that
the samples are missing the characteristic dip in the barcode distribution.
Normally we would use this dip to determine a cutoff to filter out real cells.

This may be due to there not being enough depth of sequencing or it may be
due to the samples having something wrong with them. It would be helpful to
ask the sequencing core what they think about this.

```{r barcode-setup}
samples = c("H1", "H2", "H3")
barcode_files = file.path("umis", samples, "cb-histogram.txt")
```

### K562

```{r k562-barcode}
K562_bc = read.table("metadata/k562-histogram.txt", header=FALSE, sep="\t",
                     nrows=100000)
colnames(K562_bc) = c("barcode", "count")
K562_bc$rank = as.numeric(rownames(K562_bc))
ggplot(K562_bc, aes(rank, count)) +
  scale_x_log10() +
  scale_y_log10() +
  geom_point() +
  theme_bw() +
  ggtitle("K562 barcode count distribution")
```

### K562 (60 million)

```{r k562-small-barcode}
K562_small = read.table("metadata/k562-histogram-60million.txt", header=FALSE, sep="\t",
                     nrows=100000)
colnames(K562_small) = c("barcode", "count")
K562_small$rank = as.numeric(rownames(K562_small))
ggplot(K562_small, aes(rank, count)) +
  scale_x_log10() +
  scale_y_log10() +
  geom_point() +
  theme_bw() +
  ggtitle("K562 barcode count distribution")
```

### H1
```{r h1-barcode}
h1_bc = read.table(barcode_files[1], header=FALSE, sep="\t", nrows=100000)
colnames(h1_bc) = c("barcode", "count")
h1_bc$rank = as.numeric(rownames(h1_bc))
ggplot(h1_bc, aes(rank, count)) +
  scale_x_log10() +
  scale_y_log10() +
  geom_point() +
  theme_bw() +
  ggtitle("H1 barcode count distribution")
```

### H2
```{r h2-barcode}
h2_bc = read.table(barcode_files[2], header=FALSE, sep="\t", nrows=100000)
colnames(h2_bc) = c("barcode", "count")
h2_bc$rank = as.numeric(rownames(h2_bc))
ggplot(h2_bc, aes(rank, count)) +
  scale_x_log10() +
  scale_y_log10() +
  geom_point() +
  theme_bw() +
  ggtitle("H2 barcode count distribution")
```

### H3
```{r h3-barcode}
h3_bc = read.table(barcode_files[3], header=FALSE, sep="\t", nrows=100000)
colnames(h3_bc) = c("barcode", "count")
h3_bc$rank = as.numeric(rownames(h3_bc))
ggplot(h3_bc, aes(rank, count)) +
  scale_x_log10() +
  scale_y_log10() +
  geom_point() +
  theme_bw() +
  ggtitle("H3 barcode count distribution")
```

# Allon's type barcode plots {.tabset}
This is an R port of Allon's barcode plots. The matlab command:

```
[fLog xLog] = hist(log10(counts(counts>0)),50);
y = fLog.*(10.^xLog)/sum(fLog.*(10.^xLog));
plot(xLog,y*100,'linewidth',2)
```

`hist` from matlab returns a tuple of `counts` and `centers`. R returns a
dataframe with `mids` and `counts` as columns. So, the R version of this plot is:

```{r allon-plot, echo=TRUE}
allon_barcode_plot = function(bcs, sample) {
  bcs_hist = hist(log10(bcs$count), plot=FALSE, n=50)
  fLog = bcs_hist$count
  xLog = bcs_hist$mids
  y = fLog * (10^xLog) / sum(fLog * (10^xLog))
  print(qplot(xLog, y) + geom_point() + theme_bw() + ggtitle(sample))
  return(data.frame(x=xLog, y=y, sample=sample))
}
```

This results in plots where it is easier to call a peak. It looks like we can
set a rough cutoff of around 10^4 for the H1 and H3 samples. H2 doesn't have
a clear peak but 10^4 might be a reasonable cutoff.

## K562

```{r k562-allon-plot}
k562_allon = allon_barcode_plot(K562_small, "K562")
```

## H1
```{r h1-allon-plot}
h1_allon = allon_barcode_plot(h1_bc, "H1")
```

## H2
```{r h2-allon-plot}
h2_allon = allon_barcode_plot(h2_bc, "H2")
```

## H3
```{r h3-allon-plot}
h3_allon = allon_barcode_plot(h3_bc, "H3")
```

## Combined
```{r combined-allon-plot}
vals = rbind(k562_allon, h1_allon, h2_allon, h3_allon)
ggplot(vals, aes(x, y, color=sample)) + geom_point() + theme_bw()
```

# Count based quality control
Here we compare the counts for the K562 data as well as for each of the proximal
tubule samples, using several different metrics. We counted at the level of
the transcript, so before continuing we will collapse the transcript level
counts to gene level counts by summing the counts for each transcript for each
gene.

```{r biomart-lookup}
mart = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
               dataset="mmusculus_gene_ensembl",
               host = "jul2015.archive.ensembl.org")
conversions = getBM(attributes=c("ensembl_gene_id", "refseq_mrna",
                                 "ensembl_transcript_id", "mgi_symbol"), mart=mart)
load_counts = function(count_file) {
  counts = read.table(count_file, header=TRUE, sep=",")
  counts = counts %>%
    tidyr::separate(gene, into=c("transcript", "version"), sep='\\.', fill="right") %>%
    dplyr::left_join(conversions, c("transcript"="ensembl_transcript_id")) %>%
    dplyr::filter(mgi_symbol != "") %>%
    dplyr::select(-transcript, -version, -ensembl_gene_id, -refseq_mrna) %>%
    tidyr::gather(cell, count, -mgi_symbol) %>%
    dplyr::group_by(cell, mgi_symbol) %>%
    dplyr::summarise(counts=sum(count)) %>%
    tidyr::spread(cell, counts, fill=0)
  counts = as.data.frame(counts)
  rownames(counts) = counts$mgi_symbol
  counts$mgi_symbol = NULL
  return(counts)
  }
load_bcbio_refseq = function(count_file) {
  counts = read.table(count_file, header=TRUE, sep=",")
  counts = counts %>%
    dplyr::left_join(conversions, c("gene"="refseq_mrna")) %>%
    dplyr::filter(mgi_symbol != "") %>%
    dplyr::select(-gene, -ensembl_gene_id, -ensembl_transcript_id) %>%
    tidyr::gather(cell, count, -mgi_symbol) %>%
    dplyr::group_by(cell, mgi_symbol) %>%
    dplyr::summarise(counts=sum(count)) %>%
    tidyr::spread(cell, counts, fill=0)
  counts = as.data.frame(counts)
  rownames(counts) = counts$mgi_symbol
  counts$mgi_symbol = NULL
  return(counts)
  }
```

```{r biomart-lookup-human}
mart = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
               dataset="hsapiens_gene_ensembl",
               host = "jul2015.archive.ensembl.org")
hconversions = getBM(attributes=c("ensembl_gene_id", "refseq_mrna",
                                 "ensembl_transcript_id", "hgnc_symbol"), mart=mart)
load_hcounts = function(count_file) {
  counts = read.table(count_file, header=TRUE, sep=",")
  counts = counts %>%
    dplyr::left_join(hconversions, c("gene"="ensembl_transcript_id")) %>%
    dplyr::filter(hgnc_symbol != "") %>%
    dplyr::select(-gene, -ensembl_gene_id, -refseq_mrna) %>%
    tidyr::gather(cell, count, -hgnc_symbol, -hgnc_symbol) %>%
    dplyr::group_by(cell, hgnc_symbol) %>%
    dplyr::summarise(counts=sum(count)) %>%
    tidyr::spread(cell, counts, fill=0)
  counts = as.data.frame(counts)
  rownames(counts) = counts$hgnc_symbol
  counts$hgnc_symbol = NULL
  return(counts)
  }
```

```{r k562-load-function}
load_k562 = function(count_file) {
  k562 = read.table(count_file, header=TRUE, sep=",") %>%
    tidyr::separate(gene, into=c("transcript", "symbol"), sep='\\|') %>%
    dplyr::select(-transcript) %>%
    tidyr::gather(cell, count, -symbol) %>%
    dplyr::group_by(cell, symbol) %>%
    dplyr::summarise(counts=sum(count)) %>%
    tidyr::spread(cell, counts, fill=0)
  k562 = as.data.frame(k562)
  rownames(k562) = k562$symbol
  k562$symbol = NULL
  k562 = k562[!grepl("ERCC", rownames(k562)),]
  return(k562)
}
```

```{r load-counts}
count_files = file.path("no-ncrna", samples, paste(samples, ".counts", sep=""))
h1 = load_counts(count_files[1])
h2 = load_counts(count_files[2])
h3 = load_counts(count_files[3])
count_files = file.path("new-counts", "umis", samples,
                        paste(samples, ".counts", sep=""))
h1_refseq = load_bcbio_refseq(count_files[1])
h2_refseq = load_bcbio_refseq(count_files[2])
h3_refseq = load_bcbio_refseq(count_files[3])
k562_grch37 = load_hcounts("metadata/k562-GRCh37.counts")
k562_refseq = load_k562("metadata/k562-refseq.counts")
```

Based on Allon's plots it looks like 10^4 is the cutoff for the H1-H3
samples. We already used that cutoff when we were deciding which
cells to count so we won't do any further filtering. We need to filter the
K562 samples to pick out their peak, which is more like 10^5.

```{r k562-keep}
keep_barcodes = function(barcodes, cutoff) {
  bc = subset(barcodes, count > cutoff)$barcode
  return(gsub("-", ".", bc))
}
k562_keep = keep_barcodes(K562_bc, 10^5)
k562_grch37 = k562_grch37[, k562_keep]
k562_refseq = k562_refseq[, k562_keep]
```

## Genes Detected

The H2-H3 samples have much less genes detected even when we drop
the sequencing down of the K562 samples down to 60 million reads.
The numbers of genes detected between the Ensembl and Refseq
annotations look similar.

```{r genes-detected-boxplot}
k562_grch37_detected = data.frame(cell=colnames(k562_grch37),
                                  total=colSums(k562_grch37),
                                  detected=colSums(k562_grch37 > 0),
                                  sample="k562_grch37")
k562_refseq_detected = data.frame(cell=colnames(k562_refseq),
                                  total=colSums(k562_refseq),
                                  detected=colSums(k562_refseq > 0),
                                  sample="k562_refseq")
h1_detected = data.frame(cell=colnames(h1), total=colSums(h1),
                           detected=colSums(h1 > 0), sample="h1")
h2_detected = data.frame(cell=colnames(h2), total=colSums(h2),
                           detected=colSums(h2 > 0), sample="h2")
h3_detected = data.frame(cell=colnames(h3), total=colSums(h3),
                           detected=colSums(h3 > 0), sample="h3")
h1_refseq_detected = data.frame(cell=colnames(h1_refseq), total=colSums(h1_refseq),
                           detected=colSums(h1_refseq > 0), sample="h1_refseq")
h2_refseq_detected = data.frame(cell=colnames(h2_refseq), total=colSums(h2_refseq),
                           detected=colSums(h2_refseq > 0), sample="h2_refseq")
h3_refseq_detected = data.frame(cell=colnames(h3_refseq), total=colSums(h3_refseq),
                           detected=colSums(h3_refseq > 0), sample="h3_refseq")
detected = rbind(k562_grch37_detected, h1_detected, h2_detected, h3_detected,
                 k562_refseq_detected,
                 h1_refseq_detected, h2_refseq_detected, h3_refseq_detected)
detected = transform(detected, cell=reorder(cell, -detected))

ggplot(detected, aes(sample, detected)) + geom_boxplot() +
  theme_bw()
```

Looking at histograms and the boxplots, we see about the same number of genes
detected irregardless of what annotation we use.

```{r genes-detected-histogram}
ggplot(detected, aes(reorder(cell, detected), detected)) +
  geom_bar(stat='identity', position='dodge') +
  facet_wrap(~sample, scale='free') +
  theme_bw() +
  scale_x_discrete(breaks=NULL) +
  xlab("cell") +
  ylab("genes detected")
```

## Total counts
For the K562 samples, the RefSeq annotation captures more counts per cell.
This isn't true of the H1-H3 samples, though.
```{r ensembl-vs-refseq-boxplot}
ggplot(detected, aes(sample, total)) +
  geom_boxplot() +
  theme_bw() +
  ylab("total counts per cell") +
  xlab("")
```

## Complexity
We can also try to measure the complexity of the library, looking at how many
different genes are detected plotted against the total number of counts. Here
we can see that these libraries have a fairly poor distribution of counts when
using the Ensembl annotation for H1-H3. Using the RefSeq annotation, it looks
much better. The Ensembl annotation has a bunch of non-coding RNA and
non-polyA RNA annotated, much more than the RefSeq annotation which could
be leading to the particularly poor quality using the Ensembl annotation.
As a comparison, the K562 sample, however, looks fine using either the Ensembl
or the RefSeq annotation.

We can see from these plots a poor sequence quality in the libraries, the
K562 sample has a fairly steep change in genes detected with an increased
total counts. The H1-H3 samples on the other hand have a much less steep
increase, a piece of evidence that these libraries are less complex than
the K562 lbraries. We are comparing across species here, so the comparison
isn't a great one.

```{r complexity}
ggplot(detected, aes(total, detected, color=sample)) + geom_point() +
 scale_x_log10() + facet_wrap(~sample) + theme_bw() + xlab("total counts") +
ylab("genes detected")
```

## Number of cells genes are expressed
Here we can see a steep dropoff for number of cells a gene is expressed
in the H1-H3 samples, meaning there are fewer genes detected as expressed
in every cell. The K562 samples, by comparison, have a much less sharp
falloff. This indicates that there might be poor capture of RNA from the cells,
leading to a less complex library.
```{r cells-per-gene}
k562_grch37_detected = data.frame(gene=rownames(k562_grch37),
                                  total=rowSums(k562_grch37),
                                  detected=rowSums(k562_grch37 > 0),
                                  sample="k562_grch37")
k562_refseq_detected = data.frame(gene=rownames(k562_refseq),
                                  total=rowSums(k562_refseq),
                                  detected=rowSums(k562_refseq > 0),
                                  sample="k562_refseq")
h1_detected = data.frame(gene=rownames(h1), total=rowSums(h1),
                           detected=rowSums(h1 > 0), sample="h1")
h2_detected = data.frame(gene=rownames(h2), total=rowSums(h2),
                           detected=rowSums(h2 > 0), sample="h2")
h3_detected = data.frame(gene=rownames(h3), total=rowSums(h3),
                         detected=rowSums(h3 > 0), sample="h3")
h1_refseq_detected = data.frame(gene=rownames(h1_refseq), total=rowSums(h1_refseq),
                           detected=rowSums(h1_refseq > 0), sample="h1_refseq")
h2_refseq_detected = data.frame(gene=rownames(h2_refseq), total=rowSums(h2_refseq),
                           detected=rowSums(h2_refseq > 0), sample="h2_refseq")
h3_refseq_detected = data.frame(gene=rownames(h3_refseq), total=rowSums(h3_refseq),
                           detected=rowSums(h3_refseq > 0), sample="h3_refseq")
detected = rbind(k562_grch37_detected, h1_detected, h2_detected, h3_detected,
                 k562_refseq_detected, h1_refseq_detected, h2_refseq_detected,
                 h3_refseq_detected)
detected_df = detected
detected = transform(detected, gene=reorder(gene, -detected))
detected = subset(detected, total > 100)
ggplot(detected, aes(reorder(gene, detected), detected)) +
  geom_bar(stat='identity', position='dodge') +
  facet_wrap(~sample, scale='free') +
  theme_bw() + scale_x_discrete(breaks=NULL) + xlab("genes") + ylab("cells detected")
```

The H1 samples look pretty bad-- it is interesting to see that there are
sets of genes which are detected as expressed using the Ensembl annotation
but not the RefSeq annotation in the H samples.

## Heatmaps of top 200 expressed genes {.tabset}
We can see the top 200 expressed genes have very variable expression across
the cells. There are many cells where there is no expression; the K562 cells
don't have this issue. We can see there are probably cells in the H* samples
that failed.

```{r}
top_200_heatmap = function(counts) {
  counts = as.matrix(counts)
  counts = head(counts[order(rowSums(counts), decreasing=TRUE),], 200)
  pheatmap(log(counts + 1), show_rownames=FALSE, show_colnames=FALSE)
}
```

### K562 GRCh37
```{r}
top_200_heatmap(k562_grch37)
```

### K562 RefSeq
```{r}
top_200_heatmap(k562_refseq)
```

### H1
```{r}
top_200_heatmap(h1)
```

### H2
```{r}
top_200_heatmap(h2)
```

### H3
```{r}
top_200_heatmap(h3)
```

### H1 RefSeq
```{r}
top_200_heatmap(h1_refseq)
```

### H2 RefSeq
```{r}
top_200_heatmap(h2_refseq)
```

### H3 RefSeq
```{r}
top_200_heatmap(h3_refseq)
```

## Drop worst cells
We'll drop the worst cells from the H* samples, where we define worst
as the cells with < 500 genes detected. Using 1000 as the cutoff leaves
us with tens of cells. This gets us down to hundreds.

```{r drop-lowest}
h1 = h1[, colSums(h1 > 0) > 500]
h2 = h2[, colSums(h2 > 0) > 500]
h3 = h3[, colSums(h3 > 0) > 500]
h1_refseq = h1_refseq[, colSums(h1_refseq > 0) > 500]
h2_refseq = h2_refseq[, colSums(h2_refseq > 0) > 500]
h3_refseq = h3_refseq[, colSums(h3_refseq > 0) > 500]
```

## Heatmaps of top 200 expressed genes after filtering {.tabset}
This definitely cleaned up the heatmaps, they look much nicer now. Still pretty
spotty expression considering these are the top expressed genes, though.

### H1
```{r h1-heatmap-filtered}
top_200_heatmap(h1)
```

### H2
```{r h2-heatmap-filtered}
top_200_heatmap(h2)
```

### H3
```{r h3-heatmap-filtered}
top_200_heatmap(h3)
```

### H1 RefSeq
```{r h1-refseq-heatmap-filtered}
top_200_heatmap(h1_refseq)
```

### H2 RefSeq
```{r h2-refseq-heatmap-filtered}
top_200_heatmap(h2_refseq)
```

### H3 RefSeq
```{r h3-refseq-heatmap-filtered}
top_200_heatmap(h3_refseq)
```

## Most variable genes {.tabset}
Here we take a shot at identifying the most variable genes to do PCA with,
it is tough with the samples we have because they are very low counts.

### K562 GRCh37
```{r k562-most-variable}
library(Seurat)
k562_grch37.dat = k562_grch37
colnames(k562_grch37.dat) = paste(colnames(k562_grch37.dat), "k562_grch37", sep="_")
k562_grch37.seurat = new("seurat", raw.data=log(k562_grch37.dat + 1))
k562_grch37.seurat = setup(k562_grch37.seurat, project="k562_grch37", min.cells=3, min.genes=500,
                    is.expr=0.01, names.field=2, names.delim="_")
k562_grch37.seurat=mean.var.plot(k562_grch37.seurat,y.cutoff = 2,do.plot=TRUE,x.low.cutoff=0.25,x.high.cutoff=7,fxn.x = expMean,fxn.y=logVarDivMean)
```

### H1
```{r h1-most-variable}
h1.dat = h1
colnames(h1.dat) = paste(colnames(h1.dat), "h1", sep="_")
h1.seurat = new("seurat", raw.data=log(h1.dat + 1))
h1.seurat = setup(h1.seurat, project="h1", min.cells=3, min.genes=500, is.expr=0.01,
                  names.field=2, names.delim="_")
h1.seurat=mean.var.plot(h1.seurat,y.cutoff = 2,do.plot=TRUE,x.low.cutoff=0.25,x.high.cutoff=7,fxn.x = expMean,fxn.y=logVarDivMean)
```

### H2
```{r h2-most-variable}
h2.dat = h2
colnames(h2.dat) = paste(colnames(h2.dat), "h2", sep="_")
h2.seurat = new("seurat", raw.data=log(h2.dat + 1))
h2.seurat = setup(h2.seurat, project="h2", min.cells=3, min.genes=500, is.expr=0.01,
                  names.field=2, names.delim="_")
h2.seurat=mean.var.plot(h2.seurat,y.cutoff = 2,do.plot=TRUE,x.low.cutoff=0.25,x.high.cutoff=7,fxn.x = expMean,fxn.y=logVarDivMean)
```

### H3
```{r h3-most-variable}
h3.dat = h3
colnames(h3.dat) = paste(colnames(h3.dat), "h3", sep="_")
h3.seurat = new("seurat", raw.data=log(h3.dat + 1))
h3.seurat = setup(h3.seurat, project="h3", min.cells=3, min.genes=500, is.expr=0.01,
                  names.field=2, names.delim="_")
h3.seurat=mean.var.plot(h3.seurat,y.cutoff = 2,do.plot=TRUE,x.low.cutoff=0.25,x.high.cutoff=7,fxn.x = expMean,fxn.y=logVarDivMean)
```

### H1 Refseq
```{r h1-refseq-pca}
h1_refseq.dat = h1_refseq
colnames(h1_refseq.dat) = paste(colnames(h1_refseq.dat), "h1_refseq", sep="_")
h1_refseq.seurat = new("seurat", raw.data=log(h1_refseq.dat + 1))
h1_refseq.seurat = setup(h1_refseq.seurat, project="h1_refseq", min.cells=3, min.genes=500, is.expr=0.01,
                  names.field=2, names.delim="_")
h1_refseq.seurat=mean.var.plot(h1_refseq.seurat,y.cutoff = 2,do.plot=TRUE,x.low.cutoff=0.25,x.high.cutoff=7,fxn.x = expMean,fxn.y=logVarDivMean)
```

### H2 Refseq
```{r h2-refseq-pca}
h2_refseq.dat = h2_refseq
colnames(h2_refseq.dat) = paste(colnames(h2_refseq.dat), "h2_refseq", sep="_")
h2_refseq.seurat = new("seurat", raw.data=log(h2_refseq.dat + 1))
h2_refseq.seurat = setup(h2_refseq.seurat, project="h2_refseq", min.cells=3, min.genes=500, is.expr=0.01,
                  names.field=2, names.delim="_")
h2_refseq.seurat=mean.var.plot(h2_refseq.seurat,y.cutoff = 2,do.plot=TRUE,x.low.cutoff=0.25,x.high.cutoff=7,fxn.x = expMean,fxn.y=logVarDivMean)
```

### H3 Refseq
```{r h3-refseq-pca}
h3_refseq.dat = h3_refseq
colnames(h3_refseq.dat) = paste(colnames(h3_refseq.dat), "h3_refseq", sep="_")
h3_refseq.seurat = new("seurat", raw.data=log(h3_refseq.dat + 1))
h3_refseq.seurat = setup(h3_refseq.seurat, project="h3_refseq", min.cells=3, min.genes=500, is.expr=0.01,
                  names.field=2, names.delim="_")
h3_refseq.seurat=mean.var.plot(h3_refseq.seurat,y.cutoff = 2,do.plot=TRUE,x.low.cutoff=0.25,x.high.cutoff=7,fxn.x = expMean,fxn.y=logVarDivMean)
```

## Most variable gene tables {.tabset}
### K562 GRCh37
```{r}
k562_grch37.seurat@var.genes
```

### H1
```{r}
h1.seurat@var.genes
```
### H2
```{r}
h2.seurat@var.genes
```

### H3
```{r}
h3.seurat@var.genes
```

### H1 Refseq
```{r}
h1_refseq.seurat@var.genes
```

### H2 Refseq
```{r}
h2_refseq.seurat@var.genes
```

### H3 Refseq
```{r}
h3_refseq.seurat@var.genes
```

## PCA {.tabset}
Just to see, even though the variable genes identified looks wonky,
in these samples, we ran PCA on all of the samples to see if any groups pop out.

### K562
```{r}
k562_grch37.seurat=pca(k562_grch37.seurat, pcs.print = 3, genes.print = 5, do.print=FALSE)
pca.plot(k562_grch37.seurat,pt.size = 2)
```

### H1
```{r}
h1.seurat=pca(h1.seurat, pcs.print = 3, genes.print = 5, do.print=FALSE)
pca.plot(h1.seurat,pt.size = 2)
```

### H2
```{r}
h2.seurat=pca(h2.seurat, pcs.print = 3, genes.print = 5, do.print=FALSE)
pca.plot(h2.seurat,pt.size = 2)
```

### H3
```{r}
h3.seurat=pca(h3.seurat, pcs.print = 3, genes.print = 5, do.print=FALSE)
pca.plot(h3.seurat,pt.size = 2)
```

### H1 Refseq
```{r}
h1_refseq.seurat=pca(h1_refseq.seurat, pcs.print = 3, genes.print = 5, do.print=FALSE)
pca.plot(h1_refseq.seurat,pt.size = 2)
```

### H2 Refseq
```{r}
h2_refseq.seurat=pca(h2_refseq.seurat, pcs.print = 3, genes.print = 5, do.print=FALSE)
pca.plot(h2_refseq.seurat,pt.size = 2)
```

### H3 Refseq
```{r}
h3_refseq.seurat=pca(h3_refseq.seurat, pcs.print = 3, genes.print = 5, do.print=FALSE)
pca.plot(h3_refseq.seurat,pt.size = 2)
```



## Housekeeping genes {.tabset}
I pulled a set of housekeeping genes from
[Evidence Based Selection of Housekeeping Genes](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0000898) and converted them to human and mouse symbols. Then
I made heatmaps for all of the housekeeping genes. We can see that for the K562 sample,
most of the housekeeping genes are expressed strongly whereas in the H1-H3 samples,
there is spare expression of the housekeeping genes.

```{r}
hk = as.character(read.table("metadata/hk.txt", header=FALSE)[, "V1"])
hk = c(hk, c("Actb", "Gapdh"))
human_hk = as.character(read.table("metadata/human-hk.txt", header=FALSE)[, "V1"])
human_hk = c(human_hk, c("Actb", "Gapdh"))
housekeeping_heatmap = function(counts, hk) {
  counts = counts[rownames(counts) %in% hk,]
  pheatmap(log(counts + 1), show_rownames=TRUE, show_colnames=FALSE)
}
```

### K562 GRCh37
```{r k562-grch37}
housekeeping_heatmap(k562_grch37, human_hk)
```

### K562 RefSeq
```{r k562-refseq-hk}
housekeeping_heatmap(k562_refseq, human_hk)
```

### H1
```{r h1-hk}
housekeeping_heatmap(h1, hk)
```

### H2
```{r h2-hk}
housekeeping_heatmap(h2, hk)
```

### H3
```{r h3-hk}
housekeeping_heatmap(h3, hk)
```

### H1 RefSeq
```{r h1-refseq-hk}
housekeeping_heatmap(h1_refseq, hk)
```

### H2 RefSeq
```{r h2-refseq-hk}
housekeeping_heatmap(h2_refseq, hk)
```

### H3 RefSeq
```{r h3-refseq-hk}
housekeeping_heatmap(h3_refseq, hk)
```

# Summary
It is hard to say what might be the cause of the problems we identified. If we
could see the characteristic dip in the barcode distribution like we did for
K562, I would feel pretty confident in recomending sequencing these samples some
more. Since we do not see that dip, I'm not sure what to think.

I subsampled the K562 data and reran it, the distribution looks exactly the same
and has the dip, so it isn't due to the depth of sequencing.
